java-spiffe-core/src/main/java/spiffe/workloadapi/WorkloadApiClient.java

```diff
@@ -0,0 +1,240 @@
+     * @return a {@link WorkloadApiClient}
+     */
+    public static WorkloadApiClient newClient() throws SocketEndpointAddressException {
+        val options = ClientOptions.builder().build();
+        return newClient(options);
+    }
+
+    /**
+     * Creates a new Workload API client.
+     * <p>
+     * If the SPIFFE socket endpoint address is not provided in the options, it uses the default address.
+     *
+     * @param options {@link ClientOptions}
+     * @return a {@link WorkloadApiClient}
+     */
+    public static WorkloadApiClient newClient(@NonNull ClientOptions options) throws SocketEndpointAddressException {
+        String spiffeSocketPath;
+        if (StringUtils.isNotBlank(options.spiffeSocketPath)) {
+            spiffeSocketPath = options.spiffeSocketPath;
+        } else {
+            spiffeSocketPath = Address.getDefaultAddress();
+        }
+
+        val socketEndpointAddress = Address.parseAddress(spiffeSocketPath);
+        val managedChannel = GrpcManagedChannelFactory.newChannel(socketEndpointAddress);
+
+        val workloadAPIAsyncStub = SpiffeWorkloadAPIGrpc
+                .newStub(managedChannel)
+                .withInterceptors(new SecurityHeaderInterceptor());
+
+        val workloadAPIBlockingStub = SpiffeWorkloadAPIGrpc
+                .newBlockingStub(managedChannel)
+                .withInterceptors(new SecurityHeaderInterceptor());
+
+        return new WorkloadApiClient(workloadAPIAsyncStub, workloadAPIBlockingStub, managedChannel);
+    }
+
+    /**
+     * One-shot blocking fetch call to get an X.509 context.
+     *
+     * @throws X509ContextException if there is an error fetching or processing the X.509 context
+     */
+    public X509Context fetchX509Context() {
+        Context.CancellableContext cancellableContext;
+        cancellableContext = Context.current().withCancellation();
+        X509Context result;
+        try {
+            result = cancellableContext.call(this::processX509Context);
+        } catch (Exception e) {
+            throw new X509ContextException("Error fetching X509Context", e);
+        }
+        // close connection
+        cancellableContext.close();
+        return result;
+    }
+
+    /**
+     * Watches for X.509 context updates.
+     *
+     * @param watcher an instance that implements a {@link Watcher}.
+     */
+    public void watchX509Context(Watcher<X509Context> watcher) {
+        StreamObserver<X509SVIDResponse> streamObserver = new StreamObserver<X509SVIDResponse>() {
+            @Override
+            public void onNext(X509SVIDResponse value) {
+                X509Context x509Context = null;
+                try {
+                    x509Context = GrpcConversionUtils.toX509Context(value);
+                } catch (CertificateException | X509SvidException e) {
+                    watcher.onError(new X509ContextException("Error processing X509 Context update", e));
+                }
+                watcher.onUpdate(x509Context);
```

Comment:
I believe this should be called inside the try {}, after `x509Context = GrpcConversionUtils.toX509Context(value);`. Otherwise, I believe, the onUpdate is being called when there was an error as well.

Reply:
definitely it must be inside the try block